crates/server/src/middleware/host.rs

```diff
@@ -51,73 +52,78 @@ where
     }
 
     fn call(&mut self, request: Request<Body>) -> Self::Future {
-        let result = host(request.headers(), self.listen.clone());
+        let result = host(request.headers(), &self.listen);
 
         if let Err(err) = result {
             let error_response = err.into_response();
             return Box::pin(async move { Ok(error_response) });
         }
 
-        let future = self.inner.call(request);
-
-        Box::pin(async move {
-            let response: Response = future.await?;
-            Ok(response)
-        })
+        Box::pin(self.inner.call(request))
     }
 }
 
-pub fn host(headers: &HeaderMap, listen: Vec<Multiaddr>) -> Result<(), UnauthorizedError<'static>> {
+pub fn host(headers: &HeaderMap, listen: &[Multiaddr]) -> Result<(), UnauthorizedError<'static>> {
     let caller_host = headers
         .get("referer")
         .ok_or_else(|| UnauthorizedError::new("Missing referer header"))?
         .to_str()
         .map_err(|_| UnauthorizedError::new("Invalid referer header"))?;
 
-    let ip_caller_host = normalize_origin(caller_host);
-
-    let hosts: Vec<String> = listen
-        .iter()
-        .filter_map(|addr| {
-            let mut components = addr.iter();
-            let host = match components.next() {
-                Some(multiaddr::Protocol::Ip4(host)) => host.to_string(),
-                Some(multiaddr::Protocol::Ip6(host)) => format!("[{}]", host),
-                _ => return None,
-            };
-
-            let port = match components.next() {
-                Some(multiaddr::Protocol::Tcp(port)) => port.to_string(),
-                _ => return None,
-            };
-
-            Some(format!("{}:{}", host, port))
-        })
-        .collect();
-
-    let server_host = &hosts[0];
-    if ip_caller_host == *server_host {
-        return Ok(());
+    let ip_caller_host = normalize_origin(caller_host)
+        .ok_or_else(|| UnauthorizedError::new("Invalid referer format"))?;
```

Comment:
how about split_once here?

```suggestion
    let ip_caller = normalize_origin(caller_host)
        .ok_or_else(|| UnauthorizedError::new("Invalid referer format"))?;
    let (ip_caller_host, ip_caller_port) = ip_caller.split_once(":").unwrap_or_else(|| (ip_caller, None));
```

Reply:
Updated